power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -58,29 +28,114 @@ std::tuple<Idx, Idx> count_voltage_sensors(const Idx n_bus, const MeasuredValues
     return std::make_tuple(n_voltage_sensor, n_voltage_phasor_sensor);
 }
 
-} // namespace detail
+// count flow sensors into ybus structure like
+// lower triangle part is always zero
+// for diagonal part, it will be one if there is bus injection
+// for upper triangle part, it will be one if there is branch flow sensor and the branch is fully connected
 template <symmetry_tag sym>
-inline void necessary_observability_check(MeasuredValues<sym> const& measured_values,
-                                          std::shared_ptr<MathModelTopology const> const& topo) {
+std::vector<int8_t> count_flow_sensors(MeasuredValues<sym> const& measured_values, MathModelTopology const& topo,
+                                       YBusStructure const& y_bus_structure) {
+    Idx const n_bus{topo.n_bus()};
+    std::vector<int8_t> flow_sensors(y_bus_structure.row_indptr.back(), 0); // initialize all to zero
+    for (Idx row = 0; row != n_bus; ++row) {
+        // lower triangle is ignored and kept as zero
+        // diagonal for bus injection measurement
+        if (measured_values.has_bus_injection(row)) {
+            flow_sensors[y_bus_structure.bus_entry[row]] = 1;
+        }
+        // upper triangle for branch flow measurement
+        for (Idx ybus_index = y_bus_structure.bus_entry[row] + 1; ybus_index != y_bus_structure.row_indptr[row + 1];
+             ++ybus_index) {
+            for (Idx element_index = y_bus_structure.y_bus_entry_indptr[ybus_index];
+                 element_index != y_bus_structure.y_bus_entry_indptr[ybus_index + 1]; ++element_index) {
+                YBusElement const& element = y_bus_structure.y_bus_element[element_index];
+                // shunt should not be considered
+                // if the branch is fully connected and measured, we consider it as a valid flow sensor
+                // we only need one flow sensor, so the loop will break
+                if (element.element_type != YBusElementType::shunt) {
+                    Idx const branch = element.idx;
+                    if ((measured_values.has_branch_from(branch) || measured_values.has_branch_to(branch)) &&
+                        topo.branch_bus_idx[branch][0] != -1 && topo.branch_bus_idx[branch][1] != -1) {
+                        flow_sensors[ybus_index] = 1;
```

Comment:
If we opt for `int_8` vector but one bigger,
```suggestion
                        flow_sensors[ybus_index] = 1;
                        flow_sensors.back()++;
```

Reply:
`int8` will overflow very quickly if we have more than 128 nodes. 

We can make a separate counter in this process. It can avoid one count for the necessary condition check. For the sufficient condition check, we still need to re-count.